<a id='top'></a>

# Computer Vision (Face Recognition & EmoPy Python Libraries)
## Demo
> #### IEOR 135/290, Data-X: Applied Data Ventures
> Joshua Rafael Sanchez | UC Berkeley, B.S. IEOR, '20 | joshuarafael@berkeley.edu <br>
> Made in collaboration with Ikalaq Sidhu, Arash Nourian, and Elias Castro-Hernandez.

### About This Notebook:
> This notebook contains the demo of the Computer Vision content in Data-X.  `Face_recognition` and `EmoPy` libraries are used. All questions under this demo are drawn from the concepts mentioned in the course.  

> __Instructions:__ 
> Run the cells below to run the demo.  Install the libraries before running the demo, and make sure that the libraries are up to date.  This notebook was created in Summer 2020. 

> __Resources Used:__
> - The following link goes through the content used for the `face_recognition` library in detail:  https://github.com/ageitgey/face_recognition 
> - The following link goes through the content used for the `EmoPy` library in detail:  https://github.com/thoughtworksarts/EmoPy 

> __Copyright:__ 
> Feel free to use the code as you wish.

______

<a id='table_of_contents'></a>
## Table of Contents

### [1.](#part_1)  Face Recognition (`face_recognition`)
> Here, we will import libraries first, and then go through different features which are offered in the face_recognition library.  They are divided by what you can do with the library:
- [1.0](#part_1.0) Introduction & Setup
- [1.1](#part_1.1) Finding faces in pictures (face_location)
- [1.2](#part_1.2) Find and manipulate facial features in pictures (face_landmarks)
- [1.3](#part_1.3) Recognize faces in images and identify who they are (compare_faces)
- [1.4](#part_1.4) Face recognition from webcam (video_capture)

### [2.](#part_2) EmoPy (`EmoPy`)
> We will import libraries first, and then go through an introductory example in EmoPy Python library.
- [2.0](#part_2.0) Introduction & Setup
- [2.1](#part_2.1) Generate emotion probabilites for sample image (predict)

______

<a id='part_1'></a>
## 1. Face Recognition

<a id='part_1.0'></a>
### 1.0 Introduction & Setup
[Back to the table of contents.](#table_of_contents)

#### Introduction- About the library:

> From pypi.org: <br> "Recognize and manipulate faces from Python or from the command line with
the world’s simplest face recognition library.  Built using [dlib](http://dlib.net/)’s state-of-the-art face recognition
built with deep learning. The model has an accuracy of 99.38% on the
[Labeled Faces in the Wild](http://vis-www.cs.umass.edu/lfw/) benchmark."

> **Note:** dlib is not used in this notebook.

#### Introduction- About the Files:

> The following images are used in the notebook (scaled down below). You can access them directly:
- `biden.jpg`: Portrait of Joe Biden
- `obama.jpg`: Portrait of Barack Obama
- `obama2.jpg`: Portrait of Barack Obama (different than previous)

<table><tr>
<td> <img src="biden.jpg" alt="Drawing" style="width: 200px;"/> </td>
<td> <img src="obama.jpg" alt="Drawing" style="width: 200px;"/> </td>
<td> <img src="obama2.jpg" alt="Drawing" style="width: 200px;"/> </td>
</tr></table>

#### Setup- Installing libraries (on Mac or Linux):

> First, you'll need to make sure you have the following packages installed:
- `cmake`: Free and open-source software tool
- `dlib`: Modern C++ toolkit containing machine learning algorithms and tools to create complex software

> **Tip:** First, make sure you have dlib already installed with Python bindings: [Link](https://gist.github.com/ageitgey/629d75c1baac34dfa5ca2a1928a7aeaf)

In [4]:
# You need to install dlib, probably using pip install.
# note: installing dlib might take more than a few minutes.
!pip install cmake
!pip install dlib

> Then, install `face_recognition` using pip:

In [5]:
# This may take a couple minutes.
!pip3 install face_recognition

> Then, install OpenCV. Installing may be different depending on your system.  (This will allow you to use the webcam in `face_recognition`.) <br>
> Resource: https://pypi.org/project/opencv-python/

In [34]:
# This may take 5-10 minutes, if OpenCV is not installed.
!pip3 install opencv-python

> Finally, make sure to check the version of `face_recognition`.  We used Version 1.2.3 in this notebook.

In [21]:
print('You are using Version', face_recognition.__version__)

You are using Version 1.2.3


<a id='part_1.1'></a>
### 1.1 Finding faces in pictures (`face_location`)
[Back to the table of contents.](#table_of_contents)

> Find all the faces that appear in a picture. Here is sample image input and the output:

<img src="example_1.png" alt="Drawing" style="width: 800px;"/>

> Note: We are using `biden.jpg` for this example.

> **Note:** We are importing `face_recognition` and `Image` from `PIL` here.

In [15]:
# Importing PIL.Image and face_recognition
from PIL import Image
import face_recognition

# Load the jpg file into a numpy array
image = face_recognition.load_image_file("biden.jpg") # INSERT IMAGE HERE

# Find all the faces in the image using the default HOG-based model.
# This method is fairly accurate, but not as accurate as the CNN model and not GPU accelerated.
# See also: find_faces_in_picture_cnn.py
face_locations = face_recognition.face_locations(image)

print("I found {} face(s) in this photograph.".format(len(face_locations)))

for face_location in face_locations:

    # Print the location of each face in this image
    top, right, bottom, left = face_location
    print("A face is located at pixel location Top: {}, Left: {}, Bottom: {}, Right: {}".format(top, left, bottom, right))

    # You can access the actual face itself like this:
    face_image = image[top:bottom, left:right]
    pil_image = Image.fromarray(face_image)
    pil_image.show()

I found 1 face(s) in this photograph.
A face is located at pixel location Top: 241, Left: 419, Bottom: 562, Right: 740


> In other words, the top-left corner of the image is located at (241,419), and the bottom-right corner of the image is locatd at (562, 740).

> An image like this should appear in a new pop-up:

<img src="example_1_1.png" alt="Drawing" style="width: 200px;"/>

<a id='part_1.2'></a>
### 1.2 Find and manipulate facial features in pictures (`face_landmarks`)
[Back to the table of contents.](#table_of_contents)

> Get the locations and outlines of each person's eyes, nose, mouth and chin. Here is the sample image input and the output we will be using:

<img src="example_2.png" alt="Drawing" style="width: 800px;"/>

> Note: We are using `biden.jpg` for this example. <br>
> **Note:** We are importing `face_recognition` and `Image` and `ImageDraw` from `PIL` here.
> **Tip:** Documentation for `PIL` library differs.

In [23]:
# Check the version of PIL for documentation:
import PIL # import library just to check your version.
PIL.__version__

'5.0.0'

In [6]:
from PIL import Image, ImageDraw
import face_recognition

# Load the jpg file into a numpy array
image = face_recognition.load_image_file("far_portrait_multiple.png") #INSERT IMAGE HERE

# Find all facial features in all the faces in the image
face_landmarks_list = face_recognition.face_landmarks(image)

print("I found {} face(s) in this photograph.".format(len(face_landmarks_list)))

# Create a PIL imagedraw object so we can draw on the picture
pil_image = Image.fromarray(image)
d = ImageDraw.Draw(pil_image)

for face_landmarks in face_landmarks_list:

    # Print the location of each facial feature in this image
    for facial_feature in face_landmarks.keys():
        print("The {} in this face has the following points: {}".format(facial_feature, face_landmarks[facial_feature]))

    # Let's trace out each facial feature in the image with a line!
    for facial_feature in face_landmarks.keys():
        d.line(face_landmarks[facial_feature], width=5)

# Show the picture
pil_image.show()

I found 11 face(s) in this photograph.
The chin in this face has the following points: [(1356, 440), (1356, 452), (1357, 463), (1358, 475), (1361, 486), (1368, 495), (1378, 502), (1388, 507), (1399, 509), (1410, 508), (1421, 503), (1431, 497), (1437, 488), (1441, 479), (1442, 469), (1443, 459), (1443, 449)]
The left_eyebrow in this face has the following points: [(1370, 430), (1377, 426), (1386, 425), (1394, 427), (1401, 431)]
The right_eyebrow in this face has the following points: [(1414, 433), (1421, 432), (1428, 432), (1435, 434), (1439, 440)]
The nose_bridge in this face has the following points: [(1407, 440), (1406, 446), (1406, 452), (1405, 458)]
The nose_tip in this face has the following points: [(1395, 463), (1399, 465), (1404, 466), (1408, 466), (1412, 465)]
The left_eye in this face has the following points: [(1379, 438), (1384, 436), (1389, 437), (1394, 441), (1388, 441), (1383, 440)]
The right_eye in this face has the following points: [(1416, 445), (1421, 442), (1426, 44

> You can also do really stupid stuff with applying digital make-up.  This could be the result: 

<img src="example_2_1.png" alt="Drawing" style="width: 200px;"/>

> Note: We are using `biden.jpg` for this example. <br>
> **Note:** We are importing `face_recognition` and `Image` and `ImageDraw` from `PIL` here.
> **Tip:** The following key functions are used in this sample code. 
- `polygon`: Draws a polygon.
- `line`: Draws a line between the coordinates in the `xy` list.

> Reference: Information about `ImageDraw` module. https://pillow.readthedocs.io/en/stable/reference/ImageDraw.html

In [25]:
from PIL import Image, ImageDraw
import face_recognition

# Load the jpg file into a numpy array
image = face_recognition.load_image_file("biden.jpg")

# Find all facial features in all the faces in the image
face_landmarks_list = face_recognition.face_landmarks(image)

pil_image = Image.fromarray(image)
for face_landmarks in face_landmarks_list:
    d = ImageDraw.Draw(pil_image, 'RGBA')

    # Make the eyebrows into a nightmare
    d.polygon(face_landmarks['left_eyebrow'], fill=(68, 54, 39, 128))
    d.polygon(face_landmarks['right_eyebrow'], fill=(68, 54, 39, 128))
    d.line(face_landmarks['left_eyebrow'], fill=(68, 54, 39, 150), width=5)
    d.line(face_landmarks['right_eyebrow'], fill=(68, 54, 39, 150), width=5)

    # Gloss the lips
    d.polygon(face_landmarks['top_lip'], fill=(150, 0, 0, 128))
    d.polygon(face_landmarks['bottom_lip'], fill=(150, 0, 0, 128))
    d.line(face_landmarks['top_lip'], fill=(150, 0, 0, 64), width=8)
    d.line(face_landmarks['bottom_lip'], fill=(150, 0, 0, 64), width=8)

    # Sparkle the eyes
    d.polygon(face_landmarks['left_eye'], fill=(255, 255, 255, 30))
    d.polygon(face_landmarks['right_eye'], fill=(255, 255, 255, 30))

    # Apply some eyeliner
    d.line(face_landmarks['left_eye'] + [face_landmarks['left_eye'][0]], fill=(0, 0, 0, 110), width=6)
    d.line(face_landmarks['right_eye'] + [face_landmarks['right_eye'][0]], fill=(0, 0, 0, 110), width=6)

    pil_image.show()

<a id='part_1.3'></a>
### 1.3 Recognize faces in images and identify who they are (`compare_faces`)
[Back to the table of contents.](#table_of_contents)

> Recognize who appears in each photo.

<img src="example_3.png" alt="Drawing" style="width: 800px;"/>

> We are trying to classify `obama2.jpg`, which we will define as our "unknown" image.  The function `face_encodings` should classify this image as Barack Obama. To find this out, 
- we load the JPG/PNG files into numpy arrays, 
- get the face encodings for each face in each image file, 
- and outputs the results.
<br>

> (Note: The `tolerance` is set at 0.6 because it is normally displays the best results.  Change if needed.)

In [29]:
# Load the jpg files into numpy arrays
biden_image = face_recognition.load_image_file("biden.jpg")
obama_image = face_recognition.load_image_file("obama.jpg")
unknown_image = face_recognition.load_image_file("obama2.jpg") # UNKKNOWN IMAGE

# Get the face encodings for each face in each image file
# Since there could be more than one face in each image, it returns a list of encodings.
# But since I know each image only has one face, I only care about the first encoding in each image, so I grab index 0.
try:
    biden_face_encoding = face_recognition.face_encodings(biden_image)[0]
    obama_face_encoding = face_recognition.face_encodings(obama_image)[0]
    unknown_face_encoding = face_recognition.face_encodings(unknown_image)[0]
except IndexError:
    print("I wasn't able to locate any faces in at least one of the images. Check the image files. Aborting...")
    quit()

known_faces = [
    biden_face_encoding,
    obama_face_encoding
]

# results is an array of True/False telling if the unknown face matched anyone in the known_faces array
results = face_recognition.compare_faces(known_faces, unknown_face_encoding)

print("Is the unknown face a picture of Biden? {}".format(results[0]))
print("Is the unknown face a picture of Obama? {}".format(results[1]))
print("Is the unknown face a new person that we've never seen before? {}".format(not True in results))

Is the unknown face a picture of Biden? False
Is the unknown face a picture of Obama? True
Is the unknown face a new person that we've never seen before? False


<a id='part_1.4'></a>
### 1.4 Face recognition from webcam (`video_capture`)
[Back to the table of contents.](#table_of_contents)

> You can even use this library with other Python libraries to do real-time face recognition:

<img src="example_4.gif" alt="Drawing" style="width: 800px;"/>

> **Instructions:** Run the code cell below.  After doing so, a screen should pop up showing your webcam.  If you define your face encoding in `face_encodings`, the program should work. 

> Note: The `face_recognition`, `cv2`, and `numpy` libraries are imported. <br>
> Note: Click "Python" > "Hide" to exit window.  However, the video will continue running.  So, interrupt the cell to stop the code from running.  (But the video camera will still be on.)
- Instead, click "Python" > "Exit" to quit.  However, doing so will stop Kernel, so make sure to restart Kernel. However, the video camera will turn off.

In [ ]:
# Importing appropriate libraries
import face_recognition
import cv2
import numpy as np

# This is a demo of running face recognition on live video from your webcam. It's a little more complicated than the
# other example, but it includes some basic performance tweaks to make things run a lot faster:
#   1. Process each video frame at 1/4 resolution (though still display it at full resolution)
#   2. Only detect faces in every other frame of video.

# PLEASE NOTE: This example requires OpenCV (the `cv2` library) to be installed only to read from your webcam.
# OpenCV is *not* required to use the face_recognition library. It's only required if you want to run this
# specific demo. If you have trouble installing it, try any of the other demos that don't require it instead.

# Get a reference to webcam #0 (the default one)
video_capture = cv2.VideoCapture(0)

# Load a sample picture and learn how to recognize it.
obama_image = face_recognition.load_image_file("obama.jpg")
obama_face_encoding = face_recognition.face_encodings(obama_image)[0]

# Load a second sample picture and learn how to recognize it.
biden_image = face_recognition.load_image_file("biden.jpg")
biden_face_encoding = face_recognition.face_encodings(biden_image)[0]

# ADDED: Joshua's sample picture.
joshua_image = face_recognition.load_image_file("close_portrait.png")
joshua_face_encoding = face_recognition.face_encodings(joshua_image)[0]

# Create arrays of known face encodings and their names
known_face_encodings = [
    obama_face_encoding,
    biden_face_encoding,
    joshua_face_encoding # ADDED
]
known_face_names = [
    "Barack Obama",
    "Joe Biden"
    "Joshua Sanchez" # ADDED
]

# Initialize some variables
face_locations = []
face_encodings = []
face_names = []
process_this_frame = True

while True:
    # Grab a single frame of video
    ret, frame = video_capture.read()

    # Resize frame of video to 1/4 size for faster face recognition processing
    small_frame = cv2.resize(frame, (0, 0), fx=0.25, fy=0.25)

    # Convert the image from BGR color (which OpenCV uses) to RGB color (which face_recognition uses)
    rgb_small_frame = small_frame[:, :, ::-1]

    # Only process every other frame of video to save time
    if process_this_frame:
        # Find all the faces and face encodings in the current frame of video
        face_locations = face_recognition.face_locations(rgb_small_frame)
        face_encodings = face_recognition.face_encodings(rgb_small_frame, face_locations)

        face_names = []
        for face_encoding in face_encodings:
            # See if the face is a match for the known face(s)
            matches = face_recognition.compare_faces(known_face_encodings, face_encoding)
            name = "Unknown"

            # # If a match was found in known_face_encodings, just use the first one.
            # if True in matches:
            #     first_match_index = matches.index(True)
            #     name = known_face_names[first_match_index]

            # Or instead, use the known face with the smallest distance to the new face
            face_distances = face_recognition.face_distance(known_face_encodings, face_encoding)
            best_match_index = np.argmin(face_distances)
            if matches[best_match_index]:
                name = known_face_names[best_match_index]

            face_names.append(name)

    process_this_frame = not process_this_frame


    # Display the results
    for (top, right, bottom, left), name in zip(face_locations, face_names):
        # Scale back up face locations since the frame we detected in was scaled to 1/4 size
        top *= 4
        right *= 4
        bottom *= 4
        left *= 4

        # Draw a box around the face
        cv2.rectangle(frame, (left, top), (right, bottom), (0, 0, 255), 2)

        # Draw a label with a name below the face
        cv2.rectangle(frame, (left, bottom - 35), (right, bottom), (0, 0, 255), cv2.FILLED)
        font = cv2.FONT_HERSHEY_DUPLEX
        cv2.putText(frame, name, (left + 6, bottom - 6), font, 1.0, (255, 255, 255), 1)

    # Display the resulting image
    cv2.imshow('Video', frame)

    # Hit 'q' on the keyboard to quit!
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Release handle to the webcam
video_capture.release()
cv2.destroyAllWindows()

______

<a id='part_2'></a>
## 2. EmoPy

<a id='part_2.0'></a>
### 2.0 Introduction & Setup
[Back to the table of contents.](#table_of_contents)

#### Introduction- About the library:

> EmoPy is a python toolkit with deep neural net classes which aims to make accurate predictions of emotions given images of people's faces.

> EmoPy includes several modules that are plugged together to build a trained [Facial Expression Recognition (FER)](https://en.wikipedia.org/wiki/Emotion_recognition) prediction model.
- `fermodel.py`
- `euralnets.py`
- `dataset.py`
- `data_loader.py`
- `csv_data_loader.py`
- `directory_data_loader.py`
- `data_generator.py`

#### Introduction- About the Files:

> The following images are used in the notebook (scaled down below). You can access them directly:
- `anger_image.png`: Photo of angry face.
- `disgust_image.png`: Photo of disgust face.
- `happy_image.png`: Photo of happy face.

<table><tr>
<td> <img src="anger_image.png" alt="Drawing" style="width: 200px;"/> </td>
<td> <img src="disgust_image.png" alt="Drawing" style="width: 200px;"/> </td>
<td> <img src="happy_image.png" alt="Drawing" style="width: 200px;"/> </td>
</tr></table>

#### Setup- Installing libraries:

> First, check your Python version.  EmoPy uses only Python versions 3.6 and up.

In [3]:
# Please check if you have Python 3.6 and up installed:
from platform import python_version
print(python_version())

3.6.8


> Then, make sure that you are able to download EmoPy.  This might also mean downloading Tensorflow: 
- `wrapt`: A Python module for decorators, wrappers and monkey patching.
- `Tensorflow`: TensorFlow is a free and open-source software library for dataflow and differentiable programming across a range of tasks.

In [4]:
# Below is the code we used.
!pip3 install wrapt --upgrade --ignore-installed

# This may take a couple minutes.
!pip install --upgrade tensorflow==2.0.0-beta1

# This may take a couple minutes.
!pip install EmoPy

Processing /Users/Joshua/Library/Caches/pip/wheels/32/42/7f/23cae9ff6ef66798d00dc5d659088e57dbba01566f6c60db63/wrapt-1.12.1-cp36-cp36m-macosx_10_7_x86_64.whl
Requirement already up-to-date: tensorflow==2.0.0-beta1 in /Users/Joshua/anaconda3/lib/python3.6/site-packages (2.0.0b1)


<a id='part_2.1'></a>
### 2.1 Generate emotion probabilites for sample image (`predict`)
[Back to the table of contents.](#table_of_contents)

> Import `FERModel` from `EmoPy`. 
> Note that the images are taken from the `Emopy.examples`.

In [5]:
from EmoPy.src.fermodel import FERModel
from pkg_resources import resource_filename

target_emotions = ['calm', 'anger', 'happiness']
model = FERModel(target_emotions, verbose=True)

print('Predicting on happy image...')
model.predict(resource_filename('EmoPy.examples','image_data/sample_happy_image.png'))

print('Predicting on disgust image...')
model.predict(resource_filename('EmoPy.examples','image_data/sample_disgust_image.png'))

print('Predicting on anger image...')
model.predict(resource_filename('EmoPy.examples','image_data/sample_anger_image2.png'))

/Users/Joshua/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


Initializing FER model parameters for target emotions: ['calm', 'anger', 'happiness']
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
Predicting on happy image...
anger: 44.1%
calm: 9.5%
happiness: 46.4%
Dominant emotion: happiness

Predicting on disgust image...
anger: 29.9%
calm: 16.0%
happiness: 54.1%
Dominant emotion: happiness

Predicting on anger image...
anger: 63.1%
calm: 22.9%
happiness: 14.0%
Dominant emotion: anger



______

End of content.
[Back to the table of contents.](#table_of_contents)